pandas/core/groupby/generic.py

```diff
@@ -1068,14 +1068,35 @@ def nsmallest(
         return result
 
     @doc(Series.idxmin.__doc__)
-    def idxmin(self, axis: Axis = 0, skipna: bool = True) -> Series:
-        result = self._op_via_apply("idxmin", axis=axis, skipna=skipna)
-        return result
+    def idxmin(self, axis: Axis = 0, skipna: bool = True) -> Series | DataFrame:
+        def func(df):
+            return df.idxmin(axis=axis, skipna=skipna)
+
+        func.__name__ = "idxmin"
+        return self._idxmin_idxmax(func)
 
     @doc(Series.idxmax.__doc__)
-    def idxmax(self, axis: Axis = 0, skipna: bool = True) -> Series:
-        result = self._op_via_apply("idxmax", axis=axis, skipna=skipna)
-        return result
+    def idxmax(self, axis: Axis = 0, skipna: bool = True) -> Series | DataFrame:
+        def func(df):
+            return df.idxmax(axis=axis, skipna=skipna)
+
+        func.__name__ = "idxmax"
+        return self._idxmin_idxmax(func)
+
+    def _idxmin_idxmax(self, func: Callable) -> DataFrame | Series:
```

Comment:
what if instead of all this you just did `return result.astype(self.obj.index.dtype, copy=False)`

Reply:
It seems that there are some problems with implementing it directly like this.

If it's better to make minimal changes to the code on the existing basis in this way, I will go ahead and fix these problem on this basis..

Reply:
I think we can't `dtype` directly to all `result`, because there are some types can't be converted to each other, such as `inf` and `NA`. 

Therefore, I tried to convert only the empty result.  Based on the original code, the type is incorrect only when the return value is empty.

@jbrockmendel any suggestion?